perfkitbenchmarker/scripts/execute_command.py

```diff
@@ -0,0 +1,94 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -*- coding: utf-8 -*-
+"""Runs a command, saving stdout, stderr, and the return code in files.
+
+Simplifies executing long-running commands on a remote host.
+The status file (as specified by --status) will be exclusively locked until the
+child process exits.
+To await completion, "wait_for_command.py" acquires a shared lock on the
+status file, which blocks until the process completes.
+
+*This script runs on the remote VM*.
+"""
+import fcntl
+import logging
+import optparse
+import sys
+import subprocess
+
+
+def main():
+  p = optparse.OptionParser()
+  p.add_option('-o', '--stdout', dest='stdout', metavar='FILE',
+               help="""Write stdout to FILE. Required.""")
+  p.add_option('-e', '--stderr', dest='stderr', metavar='FILE',
+               help="""Write stderr to FILE.  Required.""")
+  p.add_option('-p', '--pid', dest='pid', help="""Write PID to FILE.""",
+               metavar='FILE')
+  p.add_option('-s', '--status', dest='status', help="""Write process exit
+               status to FILE. FILE with have an exclusive lock during run.
+               Required.""", metavar='FILE')
+  p.add_option('-c', '--command', dest='command', help="""Shell command to
+               execute. Required.""")
+  options, args = p.parse_args()
+  if args:
+    print >> sys.stderr, 'Unexpected arguments: {0}'.format(args)
+    return 1
+
+  missing = []
+  for option in ('stdout', 'stderr', 'status', 'command'):
+    if getattr(options, option) is None:
+      missing.append(option)
+
+  if missing:
+    p.print_usage()
+    msg = 'Missing required flag(s): {0}'.format(
+        ', '.join('--' + i for i in missing))
+    print >> sys.stderr, msg
+    return 1
+
+  with open(options.stdout, 'w') as stdout:
+    with open(options.stderr, 'w') as stderr:
+      with open(options.status, 'w+') as status:
+        # Initialize the status to -1; will be filled with the exit status
+        # on subprocess completion.
+        status.write('-1')
+        status.flush()
+        status.seek(0)
+
+        p = subprocess.Popen(options.command, stdout=stdout, stderr=stderr,
+                             shell=True)
+        logging.info('Started pid %d: %s', p.pid, options.command)
+
+        if options.pid:
+          with open(options.pid, 'w') as pid:
+            pid.write(str(p.pid))
+
+        logging.info('Acquiring lock on %s', options.status)
+
+        fcntl.lockf(status, fcntl.LOCK_EX)
```

Comment:
Changed to nonblocking acquisition; the command will now fail if the lock cannot be acquired.
PKB generates UUIDs for each command run with this script, so the file should not exist prior to the command running. If the file cannot be locked, it's an error.